apps/src/lab2/views/SourcesContainer.tsx

```diff
@@ -56,11 +57,31 @@ const SourcesContainer: React.FC<
     () => getInitialSources(levelProperties, initialSources) || defaultSources
   );
 
+  // Gives an explicit signal to track when we are resetting sources
+  // via level change, a teacher switching between viewing different students, etc.
+  const [sourceInitializationCount, setSourceInitializationCount] = useState(0);
+
+  const [startOverProps, setStartOverProps] = useState<{
+    type: MessageType;
+    message?: string;
+  }>();
+
+  const reinitializeSources = useCallback(
```

Comment:
Curious if it would sense to fold this into `updateSources` with an optional `reinitialize` flag?